sw/device/tests/autogen/plic_all_irqs_test.c

```diff
@@ -0,0 +1,518 @@
+// static volatile dif_spi_host_irq_t spi_host_irq_serviced;
+// static volatile dif_sysrst_ctrl_irq_t sysrst_ctrl_irq_expected;
+// static volatile dif_sysrst_ctrl_irq_t sysrst_ctrl_irq_serviced;
+static volatile dif_uart_irq_t uart_irq_expected;
+static volatile dif_uart_irq_t uart_irq_serviced;
+// static volatile dif_usbdev_irq_t usbdev_irq_expected;
+// static volatile dif_usbdev_irq_t usbdev_irq_serviced;
+
+/**
+ * Provides external IRQ handling for this test.
+ *
+ * This function overrides the default external IRQ handler in
+ * `sw/device/lib/handler.h`.
+ */
+void handler_irq_external(void) {
+  // Find which interrupt fired at PLIC by claiming it.
+  dif_rv_plic_irq_id_t plic_irq_id;
+  CHECK(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_claim failed");
+
+  // Check if it is the right peripheral.
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == peripheral_expected,
+        "Interurpt from incorrect peripheral: exp = %d, obs = %d",
+        peripheral_expected, peripheral);
+
+#define PERIPHERAL_ISR(_peripheral, _handle, _plic_start_irq)                 \
+  do {                                                                        \
+    /* Correlate the interrupt fired at PLIC with the peripheral. */          \
+    dif_##_peripheral##_irq_t irq = (dif_##_peripheral##_irq_t)(              \
+        plic_irq_id - (dif_rv_plic_irq_id_t)_plic_start_irq);                 \
+    CHECK(irq == _peripheral##_irq_expected,                                  \
+          "Incorrect " #_handle " IRQ received at PLIC: exp = %d, obs = %d",  \
+          _peripheral##_irq_expected, irq);                                   \
+    _peripheral##_irq_serviced = irq;                                         \
+                                                                              \
+    /* Check if the same interrupt fired at the peripheral as well. */        \
+    /* TODO: read the whole status register to confirm none other pending. */ \
+    bool is_pending;                                                          \
+    CHECK(dif_##_peripheral##_irq_is_pending(&_handle, irq, &is_pending) ==   \
+              kDifOk,                                                         \
+          "dif_" #_peripheral "_irq_is_pending failed");                      \
+    CHECK(is_pending, "Interrupt %d fired at PLIC did not fire at " #_handle, \
+          irq);                                                               \
+                                                                              \
+    /* Clear the interrupt at the peripheral. */                              \
+    CHECK(dif_##_peripheral##_irq_acknowledge(&_handle, irq) == kDifOk,       \
+          "dif_" #_peripheral "_irq_acknowledge failed");                     \
+  } while (0)
```

Comment:
is there a particular reason this is enclosed in a `do { ... } while (0)` loop? or could we just use a `{ ... }` block instead? (sorry I may be missing something...)

Reply:
I think this is common technique to avoid some gotchas, see here: http://www.bruceblinn.com/linuxinfo/DoWhile.html